# most_abundand_OTU

This script produce four files, from COI zOTU table:
- **barcode.txt** that, per each library/sample, contains info about most abundand zOTU (COI barcode), taxonomy and bacteria presence
- **barcode.fasta**, containing most abundand Eucaryotic zOTU sequence (COI barcode) per each library/sample
- **euc5.txt**, containing information about Eucaryotic zOTU sequences that represents at least 5% of total Eucaryotic reads per library/sample
- **bac.txt**, containing information about all bacterial zOTUs in dataset

**Requirements:**
Unix environment, Python 3

**Usage:**
most_abundand_OTU.py <input file.txt>

**Example of input file:**
Input file is a final table generated by Symbiosis Evolution Group recommended [amplicon analysis workflow](https://github.com/symPiotr/amplicon_analysis_pipeline/blob/main/20201218_Wasp_COI_pipeline.txt).

[test_zotu_table_expanded.txt](https://github.com/deerhunter4/most_abundand_COI/files/6128914/test_zotu_table_expanded.txt)

OTU_ID | OTU_assignment | Taxonomy | Sequence | Total | SZF2Y9_1 | SZF2Y9_2 | SZF2Y9_3 | SZF2Y9_4
-------|----------------|----------|----------|-------|----------|----------|----------|---------
Zotu4 | otu4 | Eukaryota(1.00),Arthropoda(1.00),(...) | AATAAATAATATAAGTT(...) | 2128 | 2126 | 0 | 0 | 2	
Zotu6 | otu6 | Bacteria(1.00),Proteobacteria(1.00),(...) | TATGAATAATTTAAGTT(...) | 72 | 0 | 72 | 0 | 0	

